### PR TITLE
Added toggle button for sidebar

### PR DESCRIPTION
**Pull request by namantechlabs**
Fixed issue #70 for GCI

### CODE REVIEW

**Comment by aayush-05** [comment]
Use this :
@media only screen and (min-width: 739px) {
  #main-content.moved {
    margin-left: calc((100% - 739px) / 2 + -80px);
  }
  #headerid.moved {
    margin-left: calc((100% - 739px) / 2 + -80px);
  }

  #sidebarCollapse.moved {
    margin-left: calc((100% - 739px) / 2 + -80px);
  }
}

Rather than :
image

**Comment by namantechlabs** [comment]
made changes according to @aayush-05's comment & and sidebar toggle button is hidden when in mobile view.

**Comment by aayush-05** [comment]
Attach some screenshots and a video for the changes made in the description @namantechlabs

**Comment by namantechlabs** [comment]
Images:

**Comment by namantechlabs** [comment]
Video:
[video.zip](https://github.com/CircuitVerse/Interactive-Book/files/3936232/video.zip)

File: _includes/head.html [hunk 1]
```diff
@@ -3,7 +3,8 @@
   <meta http-equiv="X-UA-Compatible" content="IE=Edge">
 
   {% if site.plugins.jekyll-seo == nil %}
-    <title>{{ page.title }} - {{ site.title }}</title>
+ <title>{{ page.title }} - {{ site.title }}</title>
```
Comment (tachyons) [inline _includes/head.html]: Intentional change ?

File: _layouts/default.html [hunk 2]
```diff
@@ -92,5 +114,22 @@ <h2 class="text-delta">Table of contents</h2>
...
+    $('#sidebarCollapse').toggleClass('moved');
+
+  });
+
+});</script>
+
+
+
+
+
```
Comment (tachyons) [inline _layouts/default.html]: remove unnecessary blank lines

File: _sass/content.scss [hunk 3]
```diff
@@ -112,3 +112,32 @@
...
+  #sidebarCollapse.moved {
+    margin-left: calc((100% - 739px) / 2 + -80px);
+  }
+}
+
+@media only screen and (max-width: 739px) {
+  #sidebarCollapse {
+    display: none;
+  }
+
```
Comment (tachyons) [inline _sass/content.scss]: ✂️

File: _includes/head.html [hunk 4]
```diff
@@ -29,6 +30,10 @@
   <script type="text/javascript" src="{{ "/assets/js/just-the-docs.js" | absolute_url }}"></script>
 
   <meta name="viewport" content="width=device-width, initial-scale=1">
+  
+  <script defer src="https://use.fontawesome.com/releases/v5.0.13/js/solid.js" integrity="sha384-tzzSw1/Vo+0N5UhStP3bvwWPq+uvzCMfrN1fEFe+xBmv1C/AtVX5K0uZtmcHitFZ" crossorigin="anonymous"></script>
+  <script defer src="https://use.fontawesome.com/releases/v5.0.13/js/fontawesome.js" integrity="sha384-6OIrr52G08NpOFSZdxxz1xdNSndlD4vdcf/q2myIUVO0VsqaGHJsB0RaBE01VTOY" crossorigin="anonymous"></script>
```
Comment (tachyons) [inline _includes/head.html]: load assets from assets/vendor directory instead of cdn

File: _layouts/default.html [hunk 5]
```diff
@@ -5,8 +5,15 @@
 <script src="../assets/js/module.js"></script>
 <body>
 
+
+<script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
+<!-- Popper.JS -->
+<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.0/umd/popper.min.js" integrity="sha384-cs/chFZiN24E4KMATLdqdvsezGxaGsi4hLGOzlXwp5UZB1LY//20VyM2taTB4QvJ" crossorigin="anonymous"></script>
+<!-- Bootstrap JS -->
+<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.0/js/bootstrap.min.js" integrity="sha384-uefMccjFJAIv6A+rW+L4AHf99KvxDjWSu1z9VI8SKNVmz4sk7buKt/6v9KI65qnm" crossorigin="anonymous"></script>
```
Comment (tachyons) [inline _layouts/default.html]: Same as above, use assets/vendor directory to store third party libraries

File: _includes/head.html (hunk 1 above)
Comment (namantechlabs) [reply _includes/head.html]: Nope. Fixing it.

File: _layouts/default.html (hunk 2 above)
Comment (namantechlabs) [reply _layouts/default.html]: done.

File: _includes/head.html (hunk 4 above)
Comment (namantechlabs) [reply _includes/head.html]: Done. Added files in /assets/js/vendor.

File: _layouts/default.html (hunk 5 above)
Comment (namantechlabs) [reply _layouts/default.html]: Done. Added files in /assets/js/vendor.

**Comment by namantechlabs** [comment]
Hello, I have fixed all the errors. Thanks.

**Comment by aayush-05** [comment]
Please review it again @tachyons

**Comment by tachyons** [comment]
@nitin10s

**Comment by nitin10s** [comment]
1) Write just Sidebar instead of Toggle Sidebar
2) When the sidebar is active, make the button background green and text as white (basically show the active state)
Everything else looks good to me.

**Comment by namantechlabs** [comment]
Done.
image

image

File: _includes/head.html [hunk 6]
```diff
@@ -3,7 +3,8 @@
   <meta http-equiv="X-UA-Compatible" content="IE=Edge">
 
   {% if site.plugins.jekyll-seo == nil %}
-    <title>{{ page.title }} - {{ site.title }}</title>
+   <title>{{ page.title }} - {{ site.title }}</title>
+
```
Comment (tachyons) [inline _includes/head.html]: We still have unnecessary changes here

**Comment by tachyons** [comment]
Do we really need to introduce 3 java script libraries just to introduce single button ?

File: assets/js/vendor/popper.min.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
...
...upported in future versions of Popper.js!');var s,d,a=void 0===p?t.gpuAcceleration:p,l=r(e.instance.popper),f=g(l),m={position:i.position},h={left:J(i.left),top:J(i.top),bottom:J(i.bottom),right:J(i.right)},c='bottom'===o?'top':'bottom',u='right'===n?'left':'right',b=W('transform');if(d='bottom'==c?-f.height+h.bottom:h.top,s='right'==u?-f.width+h.right:h.left,a&&b)m[b]='translate3d('+s+'px, '+d+'px, 0)',m[c]=0,m[u]=0,m.willChange='transform';else{var y='bottom'==c?-1:1,w='right'==u?-1:1;m[c]=d*y,m[u]=s*w,m.willChange=c+', '+u}var E={"x-placement":e.placement};return e.attributes=de({},E,e.attributes),e.styles=de({},m,e.styles),e.arrowStyles=de({},e.offsets.arrow,e.arrowStyles),e},gpuAcceleration:!0,x:'bottom',y:'right'},applyStyle:{order:900,enabled:!0,fn:function(e){return Y(e.instance.popper,e.styles),j(e.instance.popper,e.attributes),e.arrowElement&&Object.keys(e.arrowStyles).length&&Y(e.arrowElement,e.arrowStyles),e},onLoad:function(e,t,o,n,i){var r=O(i,t,e,o.positionFixed),p=x(o.placement,r,t,e,o.modifiers.flip.boundariesElement,o.modifiers.flip.padding);return t.setAttribute('x-placement',p),Y(t,{position:o.positionFixed?'fixed':'absolute'}),o},gpuAcceleration:void 0}}},me});
```
[the remark this answers is not in the document]
Comment (namantechlabs) [reply assets/js/vendor/popper.min.js]: File Removed.

File: assets/js/vendor/bootstrapJS.min.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
...rentNode).hasClass(hn)){var r=rn(t).closest(_n)[0];r&&rn(r).find(En).addClass(cn),t.setAttribute("aria-expanded",!0)}n&&n()},i._jQueryInterface=function(n){return this.each(function(){var t=rn(this),e=t.data(sn);if(e||(e=new i(this),t.data(sn,e)),"string"==typeof n){if("undefined"==typeof e[n])throw new TypeError('No method named "'+n+'"');e[n]()}})},o(i,null,[{key:"VERSION",get:function(){return"4.1.0"}}]),i}(),rn(document).on(ln.CLICK_DATA_API,vn,function(t){t.preventDefault(),Tn._jQueryInterface.call(rn(this),"show")}),rn.fn.tab=Tn._jQueryInterface,rn.fn.tab.Constructor=Tn,rn.fn.tab.noConflict=function(){return rn.fn.tab=an,Tn._jQueryInterface},Tn);!function(t){if("undefined"==typeof t)throw new TypeError("Bootstrap's JavaScript requires jQuery. jQuery must be included before Bootstrap's JavaScript.");var e=t.fn.jquery.split(" ")[0].split(".");if(e[0]<2&&e[1]<9||1===e[0]&&9===e[1]&&e[2]<1||4<=e[0])throw new Error("Bootstrap's JavaScript requires at least jQuery v1.9.1 but less than v4.0.0")}(e),t.Util=Cn,t.Alert=In,t.Button=An,t.Carousel=Dn,t.Collapse=bn,t.Dropdown=Sn,t.Modal=wn,t.Popover=On,t.Scrollspy=kn,t.Tab=Pn,t.Tooltip=Nn,Object.defineProperty(t,"__esModule",{value:!0})});
```
[the remark this answers is not in the document]
Comment (namantechlabs) [reply assets/js/vendor/bootstrapJS.min.js]: File removed.

**Comment by namantechlabs** [comment]
> Do we really need to introduce 3 java script libraries just to introduce single button ?

Removed unnecessary libraries.

**Comment by namantechlabs** [comment]
> 1. Write just Sidebar instead of Toggle Sidebar
> 2. When the sidebar is active, make the button background green and text as white (basically show the active state)
>    Everything else looks good to me.

Demo Vid: [trim5.zip](https://github.com/CircuitVerse/Interactive-Book/files/3940288/trim5.zip)

**Comment by tachyons** [comment]
Can you make hound happy? It is complaining

**Comment by aayush-05** [comment]
Also, change the color of the button to the following when active :
background color : #026e57
text color : white

**Comment by namantechlabs** [comment]
Those issues are in the js library..

**Comment by nitin10s** [comment]
Yes you did it the opposite way, button should be set active when we can see the navbar.
Second, aayush is right use the theme green colour and change the icon + text to white (when active).

**Comment by aayush-05** [comment]
@tachyons is hound meant to be used for the compressed production level files ?

**Comment by aayush-05** [comment]
You changed the colours oppositely, I said to have green background and white text color when the button is active i.e when the sidebar is hidden

**Comment by nitin10s** [comment]
Looks good to me. Great work @namantechlabs !

**Comment by aayush-05** [comment]
Good to go for me as well. Just see if you can first declare the colour codes in a separate file and refer it here using the sass variable.
Please have a look @tachyons

**Comment by tachyons** [comment]
Please address the hound violations for the code you added, you can skip violations in third party library code

**Comment by namantechlabs** [comment]
> Please address the hound violations for the code you added, you can skip violations in third party library code

I have addressed them, except the one for creating variables for the colours.

**Comment by namantechlabs** [comment]
@nitin10s do I need to create variables for colours?

**Comment by nitin10s** [comment]
That would be great!

**Comment by namantechlabs** [comment]
@nitin10s @aayush-05 I have created variables for the button colours.

**Comment by aayush-05** [comment]
We don't need this property now, since the display is already set to none.
image

**Comment by aayush-05** [comment]
Also, please combine all the properties under 
**@media only screen and (min-width: 739px) {**
in a single packet. I can see this same line getting used two times.

**Comment by aayush-05** [comment]
Please don't overwrite code, but replace it. Also I can't see any mention of the class "new" in the script tag under the html file. I would say remove the "new" keyword and replace the new color theme in the previous classes such as sidebarCollapse.coloured, etc
image

**Comment by aayush-05** [comment]
No need of this
image

**Comment by aayush-05** [comment]
Remove the wrapper class
image

**Comment by namantechlabs** [comment]
> We don't need this property now, since the display is already set to none.
> image

This is for the sidebar, display is set to none for the button.

**Comment by aayush-05** [comment]
Also, we already have the sass variables declared. Use
dark green : $link-color
grey light : $base-button-color
Simply use the keyword "white" wherever necessary
image

**Comment by aayush-05** [comment]
> > We don't need this property now, since the display is already set to none.
> > image
> 
> This is for the sidebar, display is set to none for the button.

Then please use the display property for this. Specifying a particular dimension would break on more wider displays.

**Comment by namantechlabs** [comment]
> Please don't overwrite code, but replace it. Also I can't see any mention of the class "new" in the script tag under the html file. I would say remove the "new" keyword and replace the new color theme in the previous classes such as sidebarCollapse.coloured, etc
> image

The class new was used because the current styling of the button was done by bootstrap. The class new is being used currently.

**Comment by aayush-05** [comment]
> > Please don't overwrite code, but replace it. Also I can't see any mention of the class "new" in the script tag under the html file. I would say remove the "new" keyword and replace the new color theme in the previous classes such as sidebarCollapse.coloured, etc
> > image
> 
> The class new was used because the current styling of the button was done by bootstrap. The class new is being used currently.

Oh, okay. I missed the mention of "new" class in the html. In that case, just remove the code above it, sidebarCollapse without the new class since it is not required now.

**Comment by namantechlabs** [comment]
> > > Please don't overwrite code, but replace it. Also I can't see any mention of the class "new" in the script tag under the html file. I would say remove the "new" keyword and replace the new color theme in the previous classes such as sidebarCollapse.coloured, etc
> > > image
> > 
> > 
> > The class new was used because the current styling of the button was done by bootstrap. The class new is being used currently.
> 
> Oh, okay. I missed the mention of "new" class in the html. In that case, just remove the code above it, sidebarCollapse without the new class since it is not required now.

Done

**Comment by namantechlabs** [comment]
> rty for this. Sp



> > > We don't need this property now, since the display is already set to none.
> > > image
> > 
> > 
> > This is for the sidebar, display is set to none for the button.
> 
> Then please use the display property for this. Specifying a particular dimension would break on more wider displays.

Done.

**Comment by namantechlabs** [comment]
> Also, we already have the sass variables declared. Use
> dark green : $link-color
> grey light : $base-button-color
> Simply use the keyword "white" wherever necessary
> image

Done.

**Comment by namantechlabs** [comment]
> No need of this
> image

Done.

**Comment by namantechlabs** [comment]
> Also, please combine all the properties under
> **@media only screen and (min-width: 739px) {**
> in a single packet. I can see this same line getting used two times.

done.

File: _sass/content.scss [hunk 9]
```diff
@@ -112,3 +112,28 @@
...
   }
 }
+
+#sidebar {
+  min-width: 250px;
+  max-width: 680px;
+}
+#sidebar.active {
+  margin-left: -680px;
+}
```
[the remark this answers is not in the document]
Comment (tachyons) [reply _sass/content.scss]: Add an empty line as @hound suggested

**Comment by tachyons** [comment]
still hound issues are unattended, you can ignore  'Avoid using id selectors`, but fix rest of the issues in the code you added, also never hesitate ask for help in #coding channel in slack if you need any help in doing that
'

**Comment by namantechlabs** [comment]
@tachyons I have removed all the trailing whitespaces and added empty lines as required. You can have a look. Thanks!

File: _sass/content.scss [hunk 10]
```diff
@@ -112,3 +113,29 @@
...
+  #sidebarCollapse {
+    display: none;    
+  } }
+#sidebarCollapse.new.colored {
+  background-color: $base-button-color;
+  color: $link-color;
+}
+#sidebarCollapse.new {
+  background-color: $link-color;
+  color: white;
```
[the remark this answers is not in the document]
Comment (namantechlabs) [reply _sass/content.scss]: > Color `white` should be written in hexadecimal form as `#ffffff`
> Color literals like `white` should only be used in variable declarations; they should be referred to via variable everywhere else.
> Simply use the keyword "white" wherever necessary
@aayush-05 asked me to use keyword instead of hex

File: _sass/content.scss [hunk 11]
```diff
@@ -112,3 +111,39 @@
...
+  #sidebarCollapse {
+    display: none;
+    
+  }
+}
+
+#sidebarCollapse.new.colored {
+  background-color: $base-button-color;
+  color: $link-color;
+
```
[the remark this answers is not in the document]
Comment (namantechlabs) [reply _sass/content.scss]: done

File: _sass/content.scss [hunk 12]
```diff
@@ -112,3 +111,32 @@
...
+  }
+  #sidebarCollapse.moved {
+    margin-left: calc((100% - 739px) / 2 + -80px);
+  } }
+@media only screen and (max-width: 739px) {
+  #sidebarCollapse {
+    display: none;
+    
+  }
+}
```
[the remark this answers is not in the document]
Comment (namantechlabs) [reply _sass/content.scss]: done

File: _sass/content.scss (hunk 10 above)
[the remark this answers is not in the document]
Comment (aayush-05) [reply _sass/content.scss]: Use the sass variable $white. It is already declared in the code. @namantechlabs

**Comment by namantechlabs** [comment]
@aayush-05 I have changed the keyword to "$white".

**Comment by aayush-05** [comment]
Okay, I think it's all good to go now.
Awesome work Naman! And apologies for the redos at some points, hound is just a little too strict :p

**Comment by namantechlabs** [comment]
Thank you so much! :D

**Comment by namantechlabs** [comment]
Video Attached.

[Video_FINAL.zip](https://github.com/CircuitVerse/Interactive-Book/files/3945486/Video_FINAL.zip)

**Comment by aayush-05** [comment]
> Use this :
> @media only screen and (min-width: 739px) {
> #main-content.moved {
> margin-left: calc((100% - 739px) / 2 + -80px);
> }
> #headerid.moved {
> margin-left: calc((100% - 739px) / 2 + -80px);
> }
> 
> #sidebarCollapse.moved {
> margin-left: calc((100% - 739px) / 2 + -80px);
> }
> }
> 
> Rather than :
> image

@namantechlabs you hadn't seen the changes well for headerid.moved, you have to replace -80px in place of 65px, to keep the search bar in the middle.
Should I make another GCI task for this or make a pr for this myself @tachyons ?

**Comment by tachyons** [comment]
You can make the changes, no need to make that a GCI task

**Comment by namantechlabs** [comment]
@aayush-05, the +65 was intentional, because if I kept it -80, The header wouldn't be in the centre. 
See below. 

**With 65px**
image

**With -80px**
image

Thanks.
